### PR TITLE
ML.NET to AI and Machine Learning

### PR DESCRIPTION
**Pull request by LadyNaggaga**
update the ML.NET title

## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/whats-new/dotnet-9/overview.md](https://github.com/dotnet/docs/blob/bb3802ef14e6e1acb440cebe32b82446d5df354d/docs/core/whats-new/dotnet-9/overview.md) | [What's new in .NET 9](https://review.learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-9/overview?branch=pr-en-us-43401) |

### CODE REVIEW

**Review by luisquintanilla — APPROVED** [review]
LGTM.

File: docs/core/whats-new/dotnet-9/overview.md
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:1]: Shouldn't we also add something about what's new in AI then too?

Comment (luisquintanilla) [reply docs/core/whats-new/dotnet-9/overview.md:1]: Working on making updates. to address #43311 

Will have those in a couple of hours.

**Comment by luisquintanilla** [comment]
Added AI sections. Fixes #43311 

@tannergooding @michaelgsharp can you please review the TP / Tensors section.

@tarekgh can you please review the Tokenizers section.

**Comment by luisquintanilla** [comment]
@gewarren this should be ready for review when you get a chance. Thank you!

File: docs/core/whats-new/dotnet-9/overview.md [hunk 1]
```diff
@@ -53,9 +53,58 @@
...
+.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
+
+### Tokenizers
+
+The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.
+
+The latest release introduces significant new capabilities:
+
+- Tokenizers
+  - Byte-Level BPE
```
Comment (tarekgh) [inline docs/core/whats-new/dotnet-9/overview.md:69]: >Byte-Level BPE [](http://example.com/codeflow?start=4&length=14)

This one existed before so it is not introduced in the new release

File: docs/core/whats-new/dotnet-9/overview.md [hunk 2]
```diff
@@ -53,9 +53,58 @@
...
+  - SentencePiece
+  - WordPiece
```
Comment (tarekgh) [inline docs/core/whats-new/dotnet-9/overview.md:71]: Please list these as:

- Tiktoken
- Llama (based on SentencePiece)
- CodeGen
- Phi2 (based on CodeGen)
- WordPiece 
- Bert (based on WordPiece)

File: docs/core/whats-new/dotnet-9/overview.md [hunk 3]
```diff
@@ -53,9 +53,58 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
+
+### Tokenizers
+
+The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.
+
+The latest release introduces significant new capabilities:
+
+- Tokenizers
+  - Byte-Level BPE
```
Comment (jamesmontemagno) [inline docs/core/whats-new/dotnet-9/overview.md:69]:
```suggestion
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 4]
```diff
@@ -53,9 +53,58 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+In .NET 9, `TensorPrimitives` and the new `Tensor<T>` type expand AI capabilities by enabling efficient encoding, manipulation, and computation of multi-dimensional data.
+
+Improvements in the latest release of [System.Numerics.Tensors](https://www.nuget.org/packages/System.Numerics.Tensors/) include:
+
+#### TensorPrimitives
+
+- **Expanded Method Scope:** Increased from 40 to nearly 200 overloads, now including numerical operations similar to `Math`, `MathF`, and `INumber<T>`, but for spans of values.
+- **Performance Enhancements:** Many operations are now SIMD-optimized for better performance.
+- **Generic Overloads:** Supports any T that implements a certain interface, expanding beyond just spans of float values in .NET
```
Comment (jamesmontemagno) [inline docs/core/whats-new/dotnet-9/overview.md:89]:
```suggestion
- **Generic Overloads:** Supports any T that implements a certain interface, expanding beyond just spans of float values in .NET.
```

File: docs/core/whats-new/dotnet-9/overview.md (hunk 2 above)
Comment (jamesmontemagno) [reply docs/core/whats-new/dotnet-9/overview.md:70]: @tarekgh do you want to suggest the edits for this? I will accept them in what you have here looks good just want to make sure you get it looking lik eyou want

**Review by jamesmontemagno — APPROVED** [review]

File: docs/core/whats-new/dotnet-9/overview.md [hunk 5]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
 
 For more information, see [What's new in the SDK for .NET 9](sdk.md).
 
+## AI Building Blocks and Fundamentals
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:56]:
```suggestion
## AI building blocks and fundamentals
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 6]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData
+
+.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:60]:
```suggestion
.NET 9 introduces a unified layer of C# abstractions through the [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/) packages. These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 7]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+### Tokenizers
+
+The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:64]:
```suggestion
The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developers with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and preprocess text when working with local models.
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 8]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
-ML.NET is an open-source, cross-platform framework that enables integration of custom machine-learning models into .NET applications. The latest version, ML.NET 4.0, adds [additional tokenizer support](../../../machine-learning/whats-new/overview.md#additional-tokenizer-support) for tokenizers such as Tiktoken and models such as Llama and CodeGen. <!--Add info about `Tensor<T>` here and in what's new for ML.NET.-->
+[ML.NET](https://www.nuget.org/packages/Microsoft.ML/) is an open-source, cross-platform framework that enables integration of custom machine-learning models into .NET applications.
+
+ML.NET 4.0 brings the following improvements:
+
+- New ways to programatically configure `MLContext` options.
+- Load ONNX models as `Stream`.
+- DataFrame improvements.
+- (Experimental) TorchSharp ports of Llama and Phi family of models.
+- (Experimental) CausalLM pipeline APIs
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:106]: This should link out to https://learn.microsoft.com/en-us/dotnet/machine-learning/whats-new/overview.

File: docs/core/whats-new/dotnet-9/overview.md [hunk 9]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
 
 For more information, see [What's new in the SDK for .NET 9](sdk.md).
 
+## AI Building Blocks and Fundamentals
+
+### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData
+
+.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
+
+### Tokenizers
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:62]: I have this under [What's new in ML.NET](https://learn.microsoft.com/en-us/dotnet/machine-learning/whats-new/overview). Should it not be under ML.NET?

Comment (jamesmontemagno) [reply docs/core/whats-new/dotnet-9/overview.md:62]: oh nice... hmmmm yeah good point.... it all kinda blurs together to be honest.

File: docs/core/whats-new/dotnet-9/overview.md (hunk 8 above)
Comment (jamesmontemagno) [reply docs/core/whats-new/dotnet-9/overview.md:106]:
```suggestion
- (Experimental) CausalLM pipeline APIs

For more information, see [What's new in ML.NET](dotnet/machine-learning/whats-new/overview).
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 10]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.
+
+The latest release introduces significant new capabilities:
+
+- Tokenizers
+  - SentencePiece
```
Comment (tarekgh) [inline docs/core/whats-new/dotnet-9/overview.md:69]:
```suggest
    - Tiktoken for GPT (3, 3.5, 4, 4o, o1) and Llam3 models
    - Llama (based on SentencePiece) for Llama and Mistral models
    - CodeGen for code generation models like codegen-350M-mono
    - Phi2 (based on CodeGen) for Microsoft Phi2 model
    - WordPiece
    - Bert (based on WordPiece) for Bert supported models (e.g optimum--all-MiniLM-L6-v2)
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 11]
```diff
@@ -53,9 +53,59 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+  - WordPiece
+- Built-in tokenizers for the following models:
+  - GPT (3, 3.5, 4, 4o, o1)
+  - Llama
+  - Phi
+  - BERT
+  - CodeGen
```
Comment (tarekgh) [inline docs/core/whats-new/dotnet-9/overview.md:76]:
```suggestion
```

File: docs/core/whats-new/dotnet-9/overview.md [hunk 12]
```diff
@@ -53,9 +53,59 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+### Tokenizers
+
+The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developers with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and preprocess text when working with local models.
+
+The latest release introduces significant new capabilities:
+
+- Tokenizers
+  - SentencePiece
+  - WordPiece
```
Comment (tarekgh) [inline docs/core/whats-new/dotnet-9/overview.md:70]:
```suggestion
```

File: docs/core/whats-new/dotnet-9/overview.md (hunk 2 above)
Comment (tarekgh) [reply docs/core/whats-new/dotnet-9/overview.md:70]: I added the suggestions. Sorry for the multiple as somehow the line selections didn't work when I added the first.

File: docs/core/whats-new/dotnet-9/overview.md [hunk 13]
```diff
@@ -53,9 +53,57 @@ The .NET 9 SDK introduces _workload sets_, where all of your workloads stay at a
...
+
+ML.NET 4.0 brings the following improvements:
+
+- New ways to programatically configure `MLContext` options.
+- Load ONNX models as `Stream`.
+- DataFrame improvements.
+- (Experimental) TorchSharp ports of Llama and Phi family of models.
+- (Experimental) CausalLM pipeline APIs
+
+For more information, see [What's new in ML.NET](dotnet/machine-learning/whats-new/overview).
```
Comment (gewarren) [inline docs/core/whats-new/dotnet-9/overview.md:106]:
```suggestion
For more information, see [What's new in ML.NET](../../../machine-learning/whats-new/overview.md).
```

**Comment by jamesmontemagno** [comment]
@gewarren looks good to me for now, can update later if we need to

**Review by gewarren — APPROVED** [review]

**Pull request merged**